include/dd/NoiseFunctionality.hpp

```diff
@@ -0,0 +1,714 @@
+            package->cn.returnToCache(complexProb);
+        }
+
+        void applyDetNoiseSequential(dEdge& originalEdge, const qc::Targets& targets) {
+            dd::dEdge tmp = {};
+
+            std::array<mEdge, std::tuple_size_v<decltype(dd::dNode::e)>> idleOperation{};
+
+            // Iterate over qubits and check if the qubit had been used
+            for (auto targetQubit: targets) {
+                for (auto const& type: noiseEffects) {
+                    generateGate(idleOperation, type, targetQubit, getNoiseProbability(type, targets));
+                    tmp.p = nullptr;
+                    //Apply all noise matrices of the current noise effect
+                    for (int m = 0; m < sequentialNoiseMap.find(type)->second; m++) {
+                        auto tmp0 = package->conjugateTranspose(idleOperation[m]);
+                        auto tmp1 = package->multiply(originalEdge, dd::densityFromMatrixEdge(tmp0), 0, false);
+                        auto tmp2 = package->multiply(dd::densityFromMatrixEdge(idleOperation[m]), tmp1, 0, useDensityMatrixType);
+                        if (tmp.p == nullptr) {
+                            tmp = tmp2;
+                        } else {
+                            tmp = package->add(tmp2, tmp);
+                        }
+                    }
+                    package->incRef(tmp);
+                    dEdge::alignDensityEdge(&originalEdge);
+                    package->decRef(originalEdge);
+                    originalEdge = tmp;
+                    if (useDensityMatrixType) {
+                        dEdge::setDensityMatrixTrue(&originalEdge);
+                    }
+                }
+            }
+        }
+
+        void generateGate(std::array<mEdge, std::tuple_size_v<decltype(dd::dNode::e)>>& pointerForMatrices, dd::NoiseOperations noiseType, dd::Qubit target, double probability) {
+            std::array<dd::GateMatrix, std::tuple_size_v<decltype(dd::dNode::e)>> idleNoiseGate{};
+            dd::ComplexValue                                                      tmp = {};
+
+            switch (noiseType) {
+                case dd::phaseFlip: {
+                    tmp.r               = std::sqrt(1 - probability) * dd::complex_one.r;
+                    idleNoiseGate[0][0] = idleNoiseGate[0][3] = tmp;
+                    idleNoiseGate[0][1] = idleNoiseGate[0][2] = dd::complex_zero;
+                    tmp.r                                     = std::sqrt(probability) * dd::complex_one.r;
+                    idleNoiseGate[1][0]                       = tmp;
+                    tmp.r *= -1;
+                    idleNoiseGate[1][3] = tmp;
+                    idleNoiseGate[1][1] = idleNoiseGate[1][2] = dd::complex_zero;
+
+                    pointerForMatrices[0] = package->makeGateDD(idleNoiseGate[0], getNumberOfQubits(), target);
+                    pointerForMatrices[1] = package->makeGateDD(idleNoiseGate[1], getNumberOfQubits(), target);
```

Comment:
this is missing a comment with the matrix description.
please also check the formatting of the other comments as it seems to be quite a bit off.

Reply:
done